### PR TITLE
Cuda version check draft

### PR DESCRIPTION
**Pull request by jjwilke**

### CODE REVIEW

File: cmake/compile_tests/cuda_test.cc [hunk 1]
```diff
@@ -0,0 +1,37 @@
+#include <cuda_runtime.h>
+#include <cstdio>
+#include <Kokkos_Macros.hpp>
```
Comment (dalg24) [inline cmake/compile_tests/cuda_test.cc:3]: I would pass `KOKKOS_CUDA_ARCH_NUMBER` via a compile definition.

File: cmake/kokkos_arch.cmake [hunk 2]
```diff
@@ -440,3 +444,14 @@ FOREACH(Arch ${KOKKOS_ENABLED_ARCH_LIST})
...
+IF (KOKKOS_ENABLE_CUDA)
+  FIND_PACKAGE(CUDA)
+  IF (CUDA_FOUND)
+    ADD_EXECUTABLE(kokkos_cuda_version ${CMAKE_CURRENT_LIST_DIR}/compile_tests/cuda_test.cc)
+    TARGET_LINK_LIBRARIES(kokkos_cuda_version ${CUDA_LIBRARIES})
+    TARGET_INCLUDE_DIRECTORIES(kokkos_cuda_version PRIVATE ${KOKKOS_TOP_BUILD_DIR} ${KOKKOS_SOURCE_DIR}/core/src)
+  ELSE()
+    MESSAGE(SEND_ERROR "No CUDA is found. Validation cannot be performed")
+  ENDIF()
+  ADD_CUSTOM_TARGET(cuda_version COMMAND kokkos_cuda_version)
```
Comment (dalg24) [inline cmake/kokkos_arch.cmake:456]: The intent is the user would call `make cuda_version`?

File: cmake/compile_tests/cuda_test.cc [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    } else {
+      okay = false;
+    }
+  }
+  if (okay){
+    std::printf("Validation passed\n");
+    return 0;
+  } else {
+    std::printf("Validation failed\n");
+    return 0;
```
Comment (dalg24) [inline cmake/compile_tests/cuda_test.cc:35]: I suppose you want to exit failure.

File: cm_generate_makefile.bash [hunk 4]
```diff
@@ -244,6 +244,9 @@ do
     --no-examples)
       KOKKOS_DO_EXAMPLES=OFF
       ;;
+    --enable-examples)
+      KOKKOS_DO_EXAMPLES=ON
+      ;;
```
Comment (masterleinad) [inline cm_generate_makefile.bash:249]: I guess this is not really related to the idea proposed here.

File: scripts/testing_scripts/cm_test_all_sandia [hunk 5]
```diff
@@ -519,6 +519,7 @@ elif [ "$MACHINE" = "kokkos-dev-2" ]; then
   if [ "$SPOT_CHECK" = "True" ]; then
     # Format: (compiler module-list build-list exe-name warning-flag)
     COMPILERS=("gcc/7.3.0 $BASE_MODULE_LIST "OpenMP,Pthread" g++ $GCC_WARNING_FLAGS"
+               "gcc/8.3.0 $BASE_MODULE_LIST "OpenMP" g++ $GCC_WARNING_FLAGS"
```
Comment (masterleinad) [inline scripts/testing_scripts/cm_test_all_sandia:522]: ...and these and the ones below probably don't belong here either.

File: cmake/kokkos_suggest_cuda_arch.cmake [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  IF (CUDA_VERSION VERSION_GREATER_EQUAL 10)
+    LIST(APPEND POSSIBLE_CUDA_NAMES TURING75)
+  ENDIF()
+  IF (CUDA_VERSION VERSION_GREATER_EQUAL 9)
+    LIST(APPEND POSSIBLE_CUDA_NAMES VOLTA70;VOLTA72)
+  ENDIF()
+  IF (CUDA_VERSION VERSION_GREATER_EQUAL 8)
+    LIST(APPEND POSSIBLE_CUDA_NAMES PASCAL60;PASCAL61;PASCAL62)
+  ENDIF()
+  IF (CUDA_VERSION VERSION_GREATER_EQUAL 7)
```
Comment (masterleinad) [inline cmake/kokkos_suggest_cuda_arch.cmake:16]: You can assume that `CUDA_VERSION` is always greater than 8.

**Review by masterleinad — CHANGES_REQUESTED** [review]
There seem to be a lot of unrelated changes here. I would just merge the original pull request and discuss the changes here in a follow-up.

File: scripts/testing_scripts/cm_test_all_sandia (hunk 5 above)
Comment (jjwilke) [reply scripts/testing_scripts/cm_test_all_sandia:522]: some garbage trickled in due to different merge bases. Just ignore this.

File: cm_generate_makefile.bash (hunk 4 above)
Comment (jjwilke) [reply cm_generate_makefile.bash:249]: Yeah, some garbage trickled in due to different merge bases. Just ignore this.

File: cmake/kokkos_arch.cmake (hunk 2 above)
Comment (jjwilke) [reply cmake/kokkos_arch.cmake:456]: Correct. You would crash and say you must have a CUDA arch specified, but give directions on how to figure it out.

File: cmake/compile_tests/cuda_test.cc (hunk 3 above)
Comment (jjwilke) [reply cmake/compile_tests/cuda_test.cc:35]: Touche.

**Comment by jjwilke** [comment]
I suppose we could merge the original PR for now? At least for releases/master promotion I think many, many users building for CUDA are not going to know what to do without some prompting.
